File: benchmarks/bindings/ts/bench.ts
// TypeScript (napi) side of the binding benchmarks.
//
// Measures, for the @grepify/node host over the shared Rust engine:
//   * Boundary-crossing overhead:
//       - small sync call: detectCodeLanguage (cheapest napi round-trip)
//       - heavier sync call: matchCode (structural match; boundary + engine work)
//       - async mount round-trip: N sequential useMount of a trivial child
//   * Pipeline: mountEach over a generated corpus, cold run vs warm (memo-hit)
//     re-index latency.
//
// Writes a metrics JSON consumed by ../runner.py. Kept in sync with the Python
// (../python/bench.py) and Rust (../rust/src/main.rs) drivers.
//
// Imported by relative path to packages/node/index.ts (see the parity harness
// README for why the published entry does not exist in-repo).

import { readFileSync, mkdirSync, writeFileSync } from 'node:fs'
import { dirname } from 'node:path'
import { performance } from 'node:perf_hooks'
import {
  App,
  fn,
  matchCode,
  detectCodeLanguage,
  mountEach,
  splitTextRecursive,
  useMount,
  walkDir,
} from '../../../packages/node/index.ts'

const MATCH_SRC = 'def foo(a, b):\n    return a + b\n\ndef bar(x):\n    return x\n'
const MATCH_PATTERN = 'def \\NAME(\\(ARGS*\\)):'
const CHUNK_SIZE = 256
const LANGUAGE = 'markdown'

interface Args {
  corpus: string
  db: string
  metrics: string
  detectIters: number
  matchIters: number
  mountIters: number
}

function parseArgs(): Args {
  const argv = process.argv.slice(2)
  const map: Record<string, string> = {}
  for (let i = 0; i < argv.length; i += 2) map[argv[i].replace(/^--/, '')] = argv[i + 1]
  return {
    corpus: map.corpus,
    db: map.db,
    metrics: map.metrics,
    detectIters: Number(map['detect-iters'] ?? 500000),
    matchIters: Number(map['match-iters'] ?? 50000),
    mountIters: Number(map['mount-iters'] ?? 2000),
  }
}

function benchBoundarySync(detectIters: number, matchIters: number) {
  detectCodeLanguage('main.py')
  let t0 = performance.now()
  for (let i = 0; i < detectIters; i++) detectCodeLanguage('main.py')
  const detectNs = ((performance.now() - t0) / detectIters) * 1e6

  matchCode(MATCH_PATTERN, MATCH_SRC, 'python')
  t0 = performance.now()
  for (let i = 0; i < matchIters; i++) matchCode(MATCH_PATTERN, MATCH_SRC, 'python')
  const matchNs = ((performance.now() - t0) / matchIters) * 1e6

  return {
    detect_ns_per_call: Math.round(detectNs * 10) / 10,
    match_ns_per_call: Math.round(matchNs * 10) / 10,
    detect_iters: detectIters,
    match_iters: matchIters,
  }
}

const trivial = async (): Promise<number> => 0

const processFile = fn(
  async (absPath: string): Promise<number> => {
    // Memo key = logic hash + [absPath]; a warm re-run is a whole-child hit.
    const text = readFileSync(absPath, 'utf8')
    return splitTextRecursive(text, { chunkSize: CHUNK_SIZE, language: LANGUAGE }).length
  },
  { memo: true },
)

function totalStats(stats: { byComponent: Record<string, any> }) {
  const agg = { num_adds: 0, num_reprocesses: 0, num_unchanged: 0, num_deletes: 0 }
  for (const g of Object.values(stats.byComponent)) {
    agg.num_adds += g.numAdds
    agg.num_reprocesses += g.numReprocesses
    agg.num_unchanged += g.numUnchanged
    agg.num_deletes += g.numDeletes
  }
  return agg
}

async function drive(handle: { changed(): Promise<boolean> }): Promise<void> {
  while (await handle.changed()) {
    /* advance until terminated */
  }
}

async function main(): Promise<void> {
  const args = parseArgs()
  const result: Record<string, unknown> = { host: 'typescript' }

  result.boundary_sync = benchBoundarySync(args.detectIters, args.matchIters)

  // Async mount round-trip. Measured inside the update to exclude app startup.
  const rtApp = await App.open('bench_roundtrip', { dbPath: `${args.db}_rt` })
  const rtMain = async (iters: number): Promise<number> => {
    const t0 = performance.now()
    for (let i = 0; i < iters; i++) {
      await useMount<number>(`rt/${i}`, trivial)
    }
    return performance.now() - t0
  }
  const elapsed = await rtApp.update<number>(rtMain, { args: [args.mountIters] })
  result.boundary_async = {
    mount_us_per_op: Math.round((elapsed / args.mountIters) * 1000) / 1000,
    mount_iters: args.mountIters,
  }

  // Pipeline: cold vs warm.
  const pipeApp = await App.open('bench_pipeline', { dbPath: `${args.db}_pipe` })
  const pipelineMain = async (corpus: string): Promise<number> => {
    const files = walkDir(corpus, { recursive: true })
    const items: [string, string][] = files.map((f) => [f.key, f.path])
    const handle = await mountEach('process', processFile, items)
    await handle.ready()
    return items.length
  }

  let t0 = performance.now()
  const coldHandle = pipeApp.startUpdate(pipelineMain, { args: [args.corpus] })
  await drive(coldHandle)
  const coldMs = performance.now() - t0
  const coldStats = totalStats(coldHandle.stats())

  t0 = performance.now()
  const warmHandle = pipeApp.startUpdate(pipelineMain, { args: [args.corpus] })
  await drive(warmHandle)
  const warmMs = performance.now() - t0
  const warmStats = totalStats(warmHandle.stats())

  const nFiles = walkDir(args.corpus, { recursive: true }).length
  result.pipeline = {
    corpus_files: nFiles,
    cold_ms: Math.round(coldMs * 100) / 100,
    warm_ms: Math.round(warmMs * 100) / 100,
    cold_stats: coldStats,
    warm_stats: warmStats,
  }

  mkdirSync(dirname(args.metrics), { recursive: true })
  writeFileSync(args.metrics, JSON.stringify(result, null, 2), 'utf8')
}

main().then(
  () => process.exit(0),
  (err) => {
    console.error(err)
    process.exit(1)
  },
)

File: benchmarks/bindings/python/bench.py
"""Python (PyO3) side of the binding benchmarks.

Measures, for the Python host over the shared Rust engine:

  * Boundary-crossing overhead:
      - small sync call: detect_code_language (cheapest PyO3 round-trip)
      - heavier sync call: match_code (structural match; boundary + engine work)
      - async mount round-trip: N sequential useMount of a trivial child
  * Pipeline: mountEach over a generated corpus, cold run vs warm (memo-hit)
    re-index latency.

Writes a metrics JSON consumed by ../runner.py. Kept in sync with the TS
(../ts/bench.ts) and Rust (../rust/src/main.rs) drivers so the same workload is
measured in all three hosts.
"""

from __future__ import annotations

import argparse
import asyncio
import json
import pathlib
import time

import grepify as coco
from grepify.connectors import localfs  # noqa: F401  (kept: mirrors real imports)
from grepify.ops.code import match_code
from grepify.ops.text import RecursiveSplitter, detect_code_language

_MATCH_SRC = "def foo(a, b):\n    return a + b\n\ndef bar(x):\n    return x\n"
_MATCH_PATTERN = r"def \NAME(\(ARGS*\)):"

_splitter = RecursiveSplitter()
_CHUNK_SIZE = 256
_LANGUAGE = "markdown"


def bench_boundary_sync(detect_iters: int, match_iters: int) -> dict:
    # Warm up (first call may lazily initialize engine state).
    detect_code_language(filename="main.py")
    t0 = time.perf_counter()
    for _ in range(detect_iters):
        detect_code_language(filename="main.py")
    detect_ns = (time.perf_counter() - t0) / detect_iters * 1e9

    match_code(_MATCH_PATTERN, _MATCH_SRC, "python")
    t0 = time.perf_counter()
    for _ in range(match_iters):
        match_code(_MATCH_PATTERN, _MATCH_SRC, "python")
    match_ns = (time.perf_counter() - t0) / match_iters * 1e9

    return {
        "detect_ns_per_call": round(detect_ns, 1),
        "match_ns_per_call": round(match_ns, 1),
        "detect_iters": detect_iters,
        "match_iters": match_iters,
    }


@coco.fn
async def _trivial() -> int:
    return 0


@coco.fn
async def _roundtrip_main(iters: int) -> float:
    t0 = time.perf_counter()
    for i in range(iters):
        await coco.use_mount(coco.component_subpath("rt", str(i)), _trivial)
    return time.perf_counter() - t0


@coco.fn(memo=True)
async def _process(abspath: str) -> int:
    # Memo key is derived from `abspath` (+ logic hash); on a warm re-run the
    # whole child is a memo hit, so the read + split is skipped entirely.
    text = pathlib.Path(abspath).read_text(encoding="utf-8")
    return len(_splitter.split(text, chunk_size=_CHUNK_SIZE, language=_LANGUAGE))


@coco.fn
async def _pipeline_main(corpus: pathlib.Path) -> int:
    files = localfs.walk_dir(corpus, recursive=True)
    items = [(key, str(corpus / key)) async for key, _f in files.items()]
    handle = await coco.mount_each(coco.component_subpath("process"), _process, items)
    await handle.ready()
    return len(items)


def _total_stats(handle) -> dict:
    stats = handle.stats()
    if stats is None:
        return {}
    t = stats.total
    return {
        "num_adds": t.num_adds,
        "num_reprocesses": t.num_reprocesses,
        "num_unchanged": t.num_unchanged,
        "num_deletes": t.num_deletes,
    }


async def _run(args: argparse.Namespace) -> dict:
    result: dict = {"host": "python"}

    result["boundary_sync"] = bench_boundary_sync(args.detect_iters, args.match_iters)

    env = coco.Environment(coco.Settings.from_env(db_path=args.db))

    # Async mount round-trip.
    rt_app = coco.App(
        coco.AppConfig(name="bench_roundtrip", environment=env),
        _roundtrip_main,
        args.mount_iters,
    )
    elapsed = await rt_app.update()
    result["boundary_async"] = {
        "mount_us_per_op": round(elapsed / args.mount_iters * 1e6, 2),
        "mount_iters": args.mount_iters,
    }

    # Pipeline: cold vs warm.
    pipe_app = coco.App(
        coco.AppConfig(name="bench_pipeline", environment=env),
        _pipeline_main,
        args.corpus,
    )

    t0 = time.perf_counter()
    h_cold = pipe_app.update()
    n_files = await h_cold.result()
    cold_ms = (time.perf_counter() - t0) * 1000.0
    cold_stats = _total_stats(h_cold)

    t0 = time.perf_counter()
    h_warm = pipe_app.update()
    await h_warm.result()
    warm_ms = (time.perf_counter() - t0) * 1000.0
    warm_stats = _total_stats(h_warm)

    result["pipeline"] = {
        "corpus_files": n_files,
        "cold_ms": round(cold_ms, 2),
        "warm_ms": round(warm_ms, 2),
        "cold_stats": cold_stats,
        "warm_stats": warm_stats,
    }
    return result


def main() -> None:
    parser = argparse.ArgumentParser(description="Python binding benchmark")
    parser.add_argument("--corpus", type=pathlib.Path, required=True)
    parser.add_argument("--db", type=pathlib.Path, required=True)
    parser.add_argument("--metrics", type=pathlib.Path, required=True)
    parser.add_argument("--detect-iters", type=int, default=500_000)
    parser.add_argument("--match-iters", type=int, default=50_000)
    parser.add_argument("--mount-iters", type=int, default=2_000)
    args = parser.parse_args()

    result = asyncio.run(_run(args))
    args.metrics.parent.mkdir(parents=True, exist_ok=True)
    args.metrics.write_text(json.dumps(result, indent=2), encoding="utf-8")


if __name__ == "__main__":
    main()
